README.md
[![gitlocalized ](https://gitlocalize.com/repo/5631/whole_project/badge.svg)](https://gitlocalize.com/repo/5631/whole_project?utm_source=badge)

RPhoto is a small software aiming at the easy handling of digital camera's photos. (RPhoto is the next generation of IMPhoto). 
Its origin resides in the lack of a simple software capable of cropping photos with a constant ratio, to avoid white borders when printing.

Main features are :
  * Crop images with a constant width / height ratio (by example 4:3 for numeric photos)
  * Rotate / Flip photos
  * Lossless crop / rotate / flip operations.
  * Available under Linux and Windows.

See the user manual to have the extensive list of features.
A [FAQ](doc/FAQ.md) is also available on this github

Homepage : http://www.lprp.fr/soft/rphoto/rphoto_en.php3

# Install

See prebuilt binary packages in release tab for Windows and Debian.

PPA repository for Ubuntu : [ppa:rpeyron/ppa](https://launchpad.net/~rpeyron/+archive/ubuntu/ppa)
```
sudo apt-add-repository ppa:rpeyron/ppa
```
There is also an attempt of snap : https://snapcraft.io/rphoto 

Debian repository :
```
# Add repository
sudo echo "deb [arch=amd64]  http://www.lprp.fr/debian stable main" > /etc/apt/sources.list.d/lprp.list
# Add apt key (remi+debian@via.ecp.fr)
sudo apt-key adv --keyserver keys.gnupg.net --recv-keys 090B93891134CECB
# Install
sudo apt-get install rphoto
```

Other OSes should build the package (see below).

# Build instructions

## Dependancies

RPhoto depends on :
- wxWidgets  (https://www.wxwidgets.org/)
- libexif (http://libexif.sourceforge.net/)
- libjpeg (included)
- wxVillaLib (included and patched)

## Linux build

RPhoto is a standard autotools project :

```
./configure
make
```

You may install dependancies on debian systems by :
```
# Build dependancies
sudo apt-get install pkg-config libexif-dev libwxgtk3.0-dev libjpeg8-dev zip libtool autoconf automake autopoint 
# Run dependancies
sudo apt-get install libwxgtk3.0-0v5 libexif12 libjpeg-progs jhead
```


## Windows build

Windows Build is done with Visual Studio Community Edition. 
Please use the latest version in build/ (older vc version are not maintained)

Starting with vc2019, dependancies may be managed with vcpkg : 
```
vcpkg integrate install
vcpkg --triplet x86-windows-static install wxwidgets 
vcpkg --triplet x86-windows-static install libexif
```


# Contributing

Contributions are welcome. You may use GitHub issues tracker for issues, or GitHub Pull Requests.

# Screenshot 

![Screenshot](https://www.lprp.fr/files/2017/10/rphoto_scr.jpg)



ChangeLog
2022-08-15 20:45 remi  0.4.5-1
	* Updated to wxWidgets 3.2
	* Added Github Actions workflow for Windows build

2020-05-09 15:08 remi 0.4.5
   * Toolbar width adaptation
   * Fix bug with ListCtrlResize
   * Updated installer : admin privileges are no more required
   * Moved to Visual Studio 2019 and vcpkg for Windows build  

2019-01-06 18h45 remi  0.4.5-beta1
	* Fixed GTK3 not working wxINVERT
	* Fixed scrolling problems with wxRectTracker and wxLineTracker
	* Fixed typo in jhead error message
	* Removed debug messages
	* Updated to wxWidgets 3.1

2017-12-28 16h45 remi
    * Bundled czech translation from Pavel Borecki  (on launchpad.net)
    * Bundled russian translation from Eugene Marshal and Aleksey Kabanov (on launchpad.net)

2017-10-29 14h25 remi
	* Merged memory leaks fixes from Arnold

2017-08-15 14h25 remi
	* Pushed to GitHub

2015-09-20 15:28  remi  0.4.4
	* Released as 0.4.4

2013-06-11 15:52  remi  0.4.4alpha1
	* Added option to set the size of the handles (feature request from S. Groteloh)
	* Added option AutoCrop to crop before saving (feature request from S. Groteloh)

2013-06-08 15:28  remi  0.4.3
	* Released as 0.4.3

2013-05-20 15:33  remi  0.4.3alpha3
	* Added Default Folder and Remember Last Opened Folder option (feature request from Asen A. and Eric D.)

2013-05-20 15:33  remi  0.4.3alpha3
	* Updated libexif, jpegtran  to their latest versions
	* Added jhead dependancy (for -autorot)
	* Added AutoRot feature (not activated by default) (feature request from Asen and Max)
	* Added support for buggy JPEG with multi APP1 tags (will parse all tags for Exif and keep all duplicated while saving)
	* Fixed resize bug : fixed size did not take care of portrait / landscape orientation
	* Fixed temporary files not deleted when deleting original file
	* Fixed undo when deteling files (not undelete, but ignore ok)
	* Fixed hazardous selection in the tree when deleting files
    * Fixed size tracker does not adjust to screen size (with a smaller ratio)

2012-10-21 20:33  remi  0.4.3alpha1
	* Added Auto Select Max feature  (feature request from Eric D.)
	* Added Save Path and Save Path and Next feature  (feature request from Eric D.)
	* Fixed reinitialization of resize combo when changing of picture (submitted by Eric D.)

2012-07-28 17:33  remi  0.4.2
    * Added Mask off areas to be cropped off (feature requested by Marcus S. & Fank B.)
    * Added AutoSave Folder & Suffix (feature requested by Fank B.)
    * Added Confirmation when not save (Marcus S.)
    * Fixed tracker size when tracking

2012-05-28 16:17  remi  0.4.1
	* Added Guide Ratio feature (feature request from Wolfgang Z.)
    * Moved wxRectTracker to new wxEvtHandler (from Troels K)
        - no more background problems
        - wxRectTracker interface has slighly changed and is not backward compatible (easy changes)
        - working for wxRectRatioTracker and wxLineRatioTracker 
    * Moved to wxWidgets 2.9
    * Fixed some bugs reported by Marcus S.
        - Keyboard not functionning in ratio mode
        - Selection Handles disappear 
        - Crop border do not move when being dragged smaller
        - Readonly files
        - The 0x0+0,0 zone is too small
        - Translation notes (Free Rotation, Ratio : Free, Resize : No)
    * Updated documentation
    * Changing ratio now updates the tracked
    * Fixed behaviour while getting back to ratio from fixed size
    * Fixed default handles
    * Fixed AdjustRectMax when ratio < 0
    * Fixed empty filename save bug
    * Updated TODO list with features request from Marcus S.
 
2009-10-18 18:03  remi  0.4.0
	* Fixed MoveTo directory tree bug
	* Added Resize feature

2008-09-29 12:39  remi  0.3.2
	* Vista compatibility
	* Updated to wxWidgets 2.8
	* Full Unicode support
	* Moved to wxVillaLib 1.5
	* Fixed exif str64 encoding
	* Embedded wrjpgcom for comment editing
	* Added custom ratio lists
	* Added comment encoding support
	* Added default ratios requested by mail
	* Added uppercase extensions for Linux
	* Fixed many layout problems

2006-07-16 17:47  remi  0.3.0
	* Fixed Comments on Linux
	* Hacked wxDir for uppercase comparison
	* Move & Next
	* Take exe path for locales 
	* Fixed rectangle after rotating while using a different ratio
	* Distribution and Compilation preparation for release

2006-06-19 16:41  remi
	* Added Drag'n Drop & Commandline file

2006-05-28 23:00  remi  0.3.0-b1
	* Added Help Support 
    * Lot of BugFixes
    * QuickStart HTML documentation

2006-05-22 18:32  remi
	* Move To Folder (with Undo) 
    * Rename current file
	* Favorites Exif Tags
	* Comment Support (JPEG lossless, JPEG normal, and drivers who accept "COM") 
    * Fixed length bug in wxString64 
    * Fixed accelerators : now only ok in main area

2006-05-10 00:13  remi
	* Fixed wx2.6 compatibility
	* Undo and Redo feature
	* Fixed zoom / selection feature

2006-02-16 22:42  remi
	* EXIF List Control now resize softly and have tooltips for long texts
	* A JPEG Handler that loads/saves of custom tags (Exif, IPTC,...) 
	* wxStringBase64  support to encode custom tags

2005-11-22 00:18  remi
	* BugFixes for wxRectTracker 
    * FEATURE : RPhoto now helps to correct inclinaison !
	* Created wxLineTracker : a simple  line tracker
    * Upgrading to wxVillaLib 1.3

2005-10-30 17:19  remi
	* Added Basic inclinaison rectification feature

2005-03-05 21:11  remi   0.2.1
	* Switched to wxVillaLib 1.1 (http://sourceforge.net/projects/wxvillalib/) for wx2.5 support
    * Modularized wxmisc
    * Moved wxRectTracker in separate module

2004-12-06 19:05  remi
    * Modified GTK Hack architecture (replaced C++ multiple inheritance for interface by a "simple" host
	  sub-class)
    * Improved PaintDC GTK architecture
	* Merged Jeremie Fouche improvements.	
    * Split wxRectTracker in  wxRectTrackerRatio for ratio specific stuff 
    * Added external documentation to wxRectTracker

2004-01-17 23:50  remi  0.2.0
	* JPEG Quality handling & detection  

src/RPhoto.h
/* ****************************************************************************
 * RPhoto.h                                                                   *
 * -------------------------------------------------------------------------- *
 *                                                                            *
 * RPhoto : image tool for a quick and efficient use with digital photos      *
 * Copyright (C) 2004 - Rmi Peyronnet <remi+rphoto@via.ecp.fr>               *
 *                                                                            *
 * This program is free software; you can redistribute it and/or              *
 * modify it under the terms of the GNU General Public License                *
 * as published by the Free Software Foundation; either version 2             *
 * of the License, or (at your option) any later version.                     *
 *                                                                            *
 * This program is distributed in the hope that it will be useful,            *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
 * GNU General Public License for more details.                               *
 *                                                                            *
 * You should have received a copy of the GNU General Public License          *
 * along with this program; if not, write to the Free Software                *
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.*
 * http://www.gnu.org/copyleft/gpl.html                                       *
 * ************************************************************************** */

#define RPHOTO_NAME wxT("RPhoto")
#define RPHOTO_VERSION wxT("v0.4.6pre")
#define RPHOTO_BUILD wxT(__DATE__)
#define RPHOTO_COPYRIGHT wxT("2004-2020 - Remi Peyronnet")
#define RPHOTO_URL wxT("https://www.lprp.fr/rphoto/")

#define RT_DEFAULT_IMAGE wxT("test.jpg")

// lowercase AND uppercare to be able to see both in *nix ; but NOT for Windows (else files are seen twice)
#ifdef __WINDOWS__ 
#define RPHOTO_FILTERIMGLIST_ALL wxString(wxT("Image Files|*.jpg;*.jpeg;*.png;*.bmp;*.gif;*.tif;*.tiff;*.ico;*.ani;*.pcx;*.pnm;*.xpm"))
#define RPHOTO_FILTERIMGLIST wxString(wxT("BMP Files (*.bmp)|*.bmp|GIF Files (*.gif)|*.gif|JPEG Files (*.jpg;*.jpeg)|*.jpg;*.jpeg|PNG Files (*.png)|*.png|TIFF Files (*.tif; *.tiff)|*.tif;*.tiff"))
#else
#define RPHOTO_FILTERIMGLIST_ALL wxString(wxT("Image Files|*.jpg;*.jpeg;*.png;*.bmp;*.gif;*.tif;*.tiff;*.ico;*.ani;*.pcx;*.pnm;*.xpm;*.JPG;*.JPEG;*.PNG;*.BMP;*.GIF;*.TIF;*.TIFF;*.ICO;*.ANI;*.PCX;*.PNM;*.XPM"))
#define RPHOTO_FILTERIMGLIST wxString(wxT("BMP Files (*.bmp)|*.bmp;*.BMP|GIF Files (*.gif)|*.gif;*.GIF|JPEG Files (*.jpg;*.jpeg)|*.jpg;*.jpeg;*.JPG;*.JPEG|PNG Files (*.png)|*.png;*.PNG|TIFF Files (*.tif; *.tiff)|*.tif;*.tiff;*.TIF;*.TIFF"))
#endif
#define RPHOTO_FILTERIMGLIST_WORLD	(wxString(wxT("All files (*.*)|*.*|")) + RPHOTO_FILTERIMGLIST_ALL + wxT("|") + RPHOTO_FILTERIMGLIST)

#define RPHOTO_TEMP_PREFIX wxT("rphoto_")

#define RPHOTO_DEFAULT_EXIF_FAVORITES wxT("DateTime,ExposureTime,FocalLength,FNumber,ISOSpeedRatings,Flash,ExposureBiasValue,MaxApertureValue,Orientation,PixelXDimension,PixelYDimension,Make,Model,Copyright,ImageDescription,Contrast,Saturation,Sharpness")
#define RPHOTO_DEFAULT_FORCED_RATIO _("0:0 (Free)|4:3 (Digital Photo)|3:2 (Paper Photo)|16:9 (Digital Photo - TV|1:1  (Square)")
#define RPHOTO_DEFAULT_CUSTOM_RATIO _("13:9 (Paper Photo - German 1)|18:13 (Paper Photo - German 2)|176x240 (Fixed Size)")
#define RPHOTO_DEFAULT_FORCED_RESIZE _("Select new size|50% (Percentage)|640x480 (Width x Height)|800x0 (With ratio)|42kb (File size)")
#define RPHOTO_DEFAULT_CUSTOM_RESIZE _("800x600|1280x1024|100kb|300kb")
#define RPHOTO_DEFAULT_FORCED_GUIDE_RATIO _("0 (Free)|0.33 (Third)|1.618 (Golden Rule)")
#define RPHOTO_DEFAULT_CUSTOM_GUIDE_RATIO _("0.5 (Middle)")

#define RPHOTO_DEFAULT_COMMENT_ENCODING _("iso8859-15")

#ifdef _DEBUG
#define RPHOTO_NOT_IMPLEMENTED
#define RPHOTO_NOT_IMPLEMENTED_RET(x)
#else
#define RPHOTO_NOT_IMPLEMENTED  \
        wxMessageBox(_("This function is not yet implemented, \nor is being developped.\n"),_("Function not implemented."),wxICON_INFORMATION | wxOK ); \
        return;
#define RPHOTO_NOT_IMPLEMENTED_RET(x)  \
        wxMessageBox(_("This function is not yet implemented, \nor is being developped.\n"),_("Function not implemented."),wxICON_INFORMATION | wxOK ); \
        return(x);
#endif

class RPhotoApp : public wxApp
{
public:
	virtual bool OnInit();
    virtual void OnInitCmdLine(wxCmdLineParser& parser);
    virtual bool OnCmdLineParsed(wxCmdLineParser& parser);
};

DECLARE_APP(RPhotoApp)

// Visual Leak Detector (in project definition)
// #include <vld.h>


/*
Memory Leaks with VS :

#define _CRTDBG_MAP_ALLOC 
#include <stdlib.h> 
#include <crtdbg.h> 

_CrtDumpMemoryLeaks(); // must be after all clean up... where in wxWidgets ?!
*/